src/protocol/sort/bit_permutation.rs

```diff
@@ -3,83 +3,93 @@ use std::iter::{repeat, zip};
 use crate::{
     error::BoxError,
     ff::Field,
-    protocol::{context::ProtocolContext, RecordId},
-    secret_sharing::Replicated,
+    protocol::{context::ProtocolContext, RecordId, RECORD_0},
+    secret_sharing::{MaliciousReplicated, Replicated, SecretSharing},
 };
 
 use crate::protocol::mul::SecureMul;
 use embed_doc_image::embed_doc_image;
 use futures::future::try_join_all;
 
+pub trait ShareOfOne<F: Field> {
+    type Share: SecretSharing<F>;
+    fn share_of_one(&self) -> Self::Share;
+}
+
+impl<F: Field> ShareOfOne<F> for ProtocolContext<'_, Replicated<F>, F> {
+    type Share = Replicated<F>;
+
+    fn share_of_one(&self) -> Self::Share {
+        Replicated::one(self.role())
+    }
+}
+
+impl<F: Field> ShareOfOne<F> for ProtocolContext<'_, MaliciousReplicated<F>, F> {
+    type Share = MaliciousReplicated<F>;
+
+    fn share_of_one(&self) -> Self::Share {
+        MaliciousReplicated::one(self.role(), self.prss().generate_replicated(RECORD_0))
```

Comment:
@benjaminsavage @akoshelev is this the correct way to generate r_share? Context does not hold r_share and thats why I did it like this

Reply:
No, this is not correct.

`SecurityValidator` is the thing that knows what the `r_share` is. It has an accessor: `validator.r_share()`.

When the `SecurityValidator` is created via a call to `SecurityValidator::new()` that's when the global value of `r_share` is initialized. 

If you look at the tests in `malicious.rs` like `simplest_circuit` you'll see that right now nobody holds a reference to this thing... that's a problem.

Right now, I just get the `SecurityValidatorAccumulator` from it via a call to `v.accumulator()`, and this object is passed to the context via a call to `upgrade_to_malicious(acc.clone())`

Then, nobody has a reference to the actual `SecurityValidator` ever again... eventually someone needs it in order to call `validate()`

Here is what I'd suggest:

Update the `upgrade_to_malicious()` function to *also* take a clone of `r_share`. Once it has this, you can expose a new `get_share_of_one()` method on the context. Then, the context (who knows both the `role` and `r_share`) will be capable of generating a share of one in the malicious context.

Reply:
Makes complete sense! Making the change